[dynamo][hop] Introduce FlexAttentionBackwardHighOrderVariable

xmfan opened the pull request:
FIXES https://github.com/pytorch/pytorch/issues/143180

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #144533

This PR adds a new variable mapping to SourcelessBuilder to represent the flex attention intermediates. The variable proxies a call to HOP, and carryovers the graph state (subgraphs represented as UnspecializedNNModuleVariable) to the dynamo output graph. This is safe to do because the nn modules used in flex attention have either been speculated on before, or are outputs of make_fx of the forward.

tlparse of `TestCompiledAutograd.test_flex_attention`: https://manifold.edge.x2p.facebook.net/v0/read/tree/logs/.tmpiWendk/index.html?bucketName=tlparse_reports&apiKey=tlparse_reports-key&withPayload=1&timeoutMsec=100

```python
class GraphModule(torch.nn.Module):
    def forward(self, L_inputs_ : list):
         ...
         # File: /data/users/xmfan/core/b/pytorch/torch/_dynamo/compiled_autograd.py:832 in set_node_origin, code: CompiledFunctionBackward0 (NodeCall 1)
        ...
        fw_graph0_0 = self.fw_graph0_0
        joint_graph0_0 = self.joint_graph0_0
        mask_graph0_0 = self.mask_graph0_0
        flex_attention_backward = torch.ops.higher_order.flex_attention_backward(aot0_primals_1, aot0_primals_1, aot0_primals_1, aot0_detach_3, aot0_detach_5, aot0_expand_5, aot0_zeros_1, fw_graph0_0, joint_graph0_0, (1, 1, aot0_ones, aot0_zeros, None, None, aot0__to_copy_1, aot0__to_copy_2, None, None, 1073741824, 1073741824, mask_graph0_0), 0.125, {'PRESCALE_QK': False, 'ROWS_GUARANTEED_SAFE': False, 'BLOCKS_ARE_CONTIGUOUS': False, 'WRITE_DQ': True, 'OUTPUT_LOGSUMEXP': True}, (), ());  aot0_primals_1 = aot0_detach_3 = aot0_detach_5 = aot0_expand_5 = aot0_zeros_1 = fw_graph0_0 = joint_graph0_0 = aot0_ones = aot0_zeros = aot0__to_copy_1 = aot0__to_copy_2 = mask_graph0_0 = None
        aot0_getitem_4: "bf16[1, 1, s0, s1][s0*s1, s0*s1, s1, 1]cuda:0" = flex_attention_backward[0]
        aot0_getitem_5: "bf16[1, 1, s0, s1][s0*s1, s0*s1, s1, 1]cuda:0" = flex_attention_backward[1]
        aot0_getitem_6: "bf16[1, 1, s0, s1][s0*s1, s0*s1, s1, 1]cuda:0" = flex_attention_backward[2];  flex_attention_backward = None
        ...
        
    class fw_graph0_0(torch.nn.Module):
        def forward(self, arg0_1: "bf16[][]cuda:0", arg1_1: "i32[][]cuda:0", arg2_1: "i32[][]cuda:0", arg3_1: "i32[][]cuda:0", arg4_1: "i32[][]cuda:0"):
            return arg0_1
            
    class joint_graph0_0(torch.nn.Module):
        def forward(self, arg0_1: "bf16[][]cuda:0", arg1_1: "i32[][]cuda:0", arg2_1: "i32[][]cuda:0", arg3_1: "i32[][]cuda:0", arg4_1: "i32[][]cuda:0", arg5_1: "bf16[][]cuda:0"):
            return [arg5_1, None, None, None, None]
            
    class mask_graph0_0(torch.nn.Module):
        def forward(self, arg0_1: "i32[][]cuda:0", arg1_1: "i32[][]cuda:0", arg2_1: "i32[][]cuda:0", arg3_1: "i32[][]cuda:0"):
             # File: /data/users/xmfan/core/b/pytorch/torch/_dynamo/compiled_autograd.py:832 in set_node_origin, code: CompiledFunctionBackward0 (NodeCall 1)
            new_ones: "b8[][]cuda:0" = torch.ops.aten.new_ones.default(arg0_1, [], dtype = torch.bool, device = device(type='cuda', index=0), pin_memory = False);  arg0_1 = None
            return new_ones
            
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov @zou3519 @ydwu4 @Chillee @drisspg @yanboliang @BoyuanFeng

xmfan committed:
Update

[ghstack-poisoned]

zou3519 commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -2170,14 +2171,40 @@ def call_function(
 
 
-class AutoFunctionalizeHigherOrderVariable(TorchHigherOrderOperatorVariable):
+class PassthroughHigherOrderVariable(TorchHigherOrderOperatorVariable):
```
@xmfan there's two ways to go around this:
1) We can create some sort of GenericHOPVariable that works for simple HOPs. 
2) We can create a FlexAttentionBackwardHOPVariable for just FlexAttentionBackward.

2 is a lot simpler to do. If you want the task of doing 1 because it is more generic, there are some things missing in this PR (that I'll comment on inline)

zou3519 commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -2170,14 +2171,40 @@ def call_function(
-class AutoFunctionalizeHigherOrderVariable(TorchHigherOrderOperatorVariable):
+class PassthroughHigherOrderVariable(TorchHigherOrderOperatorVariable):
+    """
+    A Variable to proxy a call to the HOP into the Dynamo output graph.
+    Used by Compiled Autograd to let backward HOPs survive until AOTDispatcher.
```
this is not just used by compiled autograd. Someone can use torch.ops.higher_order.auto_functionalized in a compiled program and it will go into this Variable. Ditto with flex_attention_backward.

Can you check the following?
1) that there is a good error for input type validation. It's only OK to create this if any function inputs are OpOverload, GraphModules, or other proxy-able types. Arbitrary functions are not allowed. Otherwise, this should be an error (maybe graph break?)

xmfan replied:
This should be covered, assuming all non proxy-able types throw when `arg.as_proxy()` is called

zou3519 replied:
Can we add tests to make sure the error messages are good and that the errors are being thrown in Dynamo (and not later down the line)?

zou3519 commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -2170,14 +2171,40 @@ def call_function(
+    """
+
+    def proxy_submod(self, tx, arg):
+        assert isinstance(arg.source, DictGetItemSource)
```
This is assuming we're coming from compiled autograd tracing, which may not always be the case

ydwu4 commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -2170,14 +2171,40 @@ def call_function(
 
-        p_args = tuple(arg.as_proxy() for arg in args)
-        p_kwargs = {key: arg.as_proxy() for key, arg in kwargs.items()}
+        p_args = tuple(self.to_proxy(tx, arg) for arg in args)
```
For other hops, we usually call `speculate_subgraph` to 1. lift the closures (symints, tensors) and the free symints in tensor input as graph inputs 2. detect side-effect or something. Is there a case where flex attention backward can be called directly (i'm less worried about auto_functionalized because it takes operators)? Do we want to also have these safety checks here?

zou3519 replied:
I believe the function passed to FlexAttentionBackward is a GraphModule so there's no need to speculate subgraph and it doesn't have side effects. But it sounds like we should include some sort of alias+mutation check.

xmfan committed:
Update

[ghstack-poisoned]

xmfan commented:
Okay, I'm going back to making a variable only for flex attention backward hops. Allowing proxying of arbitrary subgraphs is annoying because the subgraph input construction logic is spread out all over the place. When subgraphs aren't involved, it's easier, and we can consolidate a few of the HOP variables in the file.

zou3519 commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -2171,6 +2172,11 @@ def call_function(
 
 class AutoFunctionalizeHigherOrderVariable(TorchHigherOrderOperatorVariable):
+    """
+    A Variable to proxy a call to the HOP into the Dynamo output graph.
+    Used by Compiled Autograd to let backward HOPs survive until AOTDispatcher.
```
@xmfan remove the comment?

xmfan committed:
Update on "[dynamo][hop] Introduce FlexAttentionBackwardHighOrderVariable"

FIXES https://github.com/pytorch/pytorch/issues/143180


This PR adds a new variable mapping to SourcelessBuilder to represent the flex attention intermediates. The variable proxies a call to HOP, and carryovers the graph state (subgraphs represented as UnspecializedNNModuleVariable) to the dynamo output graph. This is safe to do because the nn modules used in flex attention have either been speculated on before, or are outputs of make_fx of the forward.

tlparse of `TestCompiledAutograd.test_flex_attention`: https://manifold.edge.x2p.facebook.net/v0/read/tree/logs/.tmpiWendk/index.html?bucketName=tlparse_reports&apiKey=tlparse_reports-key&withPayload=1&timeoutMsec=100

```python
class GraphModule(torch.nn.Module):
    def forward(self, L_inputs_ : list):
         ...
         # File: /data/users/xmfan/core/b/pytorch/torch/_dynamo/compiled_autograd.py:832 in set_node_origin, code: CompiledFunctionBackward0 (NodeCall 1)
        ...
        fw_graph0_0 = self.fw_graph0_0
        joint_graph0_0 = self.joint_graph0_0
        mask_graph0_0 = self.mask_graph0_0
        flex_attention_backward = torch.ops.higher_order.flex_attention_backward(aot0_primals_1, aot0_primals_1, aot0_primals_1, aot0_detach_3, aot0_detach_5, aot0_expand_5, aot0_zeros_1, fw_graph0_0, joint_graph0_0, (1, 1, aot0_ones, aot0_zeros, None, None, aot0__to_copy_1, aot0__to_copy_2, None, None, 1073741824, 1073741824, mask_graph0_0), 0.125, {'PRESCALE_QK': False, 'ROWS_GUARANTEED_SAFE': False, 'BLOCKS_ARE_CONTIGUOUS': False, 'WRITE_DQ': True, 'OUTPUT_LOGSUMEXP': True}, (), ());  aot0_primals_1 = aot0_detach_3 = aot0_detach_5 = aot0_expand_5 = aot0_zeros_1 = fw_graph0_0 = joint_graph0_0 = aot0_ones = aot0_zeros = aot0__to_copy_1 = aot0__to_copy_2 = mask_graph0_0 = None
        aot0_getitem_4: "bf16[1, 1, s0, s1][s0*s1, s0*s1, s1, 1]cuda:0" = flex_attention_backward[0]
        aot0_getitem_5: "bf16[1, 1, s0, s1][s0*s1, s0*s1, s1, 1]cuda:0" = flex_attention_backward[1]
        aot0_getitem_6: "bf16[1, 1, s0, s1][s0*s1, s0*s1, s1, 1]cuda:0" = flex_attention_backward[2];  flex_attention_backward = None
        ...
        
    class fw_graph0_0(torch.nn.Module):
        def forward(self, arg0_1: "bf16[][]cuda:0", arg1_1: "i32[][]cuda:0", arg2_1: "i32[][]cuda:0", arg3_1: "i32[][]cuda:0", arg4_1: "i32[][]cuda:0"):
            return arg0_1
            
    class joint_graph0_0(torch.nn.Module):
        def forward(self, arg0_1: "bf16[][]cuda:0", arg1_1: "i32[][]cuda:0", arg2_1: "i32[][]cuda:0", arg3_1: "i32[][]cuda:0", arg4_1: "i32[][]cuda:0", arg5_1: "bf16[][]cuda:0"):
            return [arg5_1, None, None, None, None]
            
    class mask_graph0_0(torch.nn.Module):
        def forward(self, arg0_1: "i32[][]cuda:0", arg1_1: "i32[][]cuda:0", arg2_1: "i32[][]cuda:0", arg3_1: "i32[][]cuda:0"):
             # File: /data/users/xmfan/core/b/pytorch/torch/_dynamo/compiled_autograd.py:832 in set_node_origin, code: CompiledFunctionBackward0 (NodeCall 1)
            new_ones: "b8[][]cuda:0" = torch.ops.aten.new_ones.default(arg0_1, [], dtype = torch.bool, device = device(type='cuda', index=0), pin_memory = False);  arg0_1 = None
            return new_ones
            
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov zou3519 ydwu4 Chillee drisspg yanboliang BoyuanFeng

[ghstack-poisoned]

xmfan commented:
@pytorchbot merge

xmfan commented:
@pytorchbot merge

huydhn commented:
This seems to cause an OOM on ROCm.  I'm not sure if this is due to the change or due to the new MI300 runners though.

inductor/test_compiled_autograd.py::TestCompiledAutograd::test_flex_attention [GH job link](https://github.com/pytorch/pytorch/actions/runs/12794902122/job/35676783114) [HUD commit link](https://hud.pytorch.org/pytorch/pytorch/commit/898a90c6bb9d1e605e6d1e8e066455330e0c653b)

cc @xmfan @jeffdaily @jithunnair-amd
